tests/e2e/watch_test.go

```diff
@@ -311,24 +315,370 @@ func TestDeleteEventDrop_Issue18089(t *testing.T) {
-		}
-		if watchResp.Events[0].Type != mvccpb.PUT {
-			t.Fatalf("unexpected event type, expected: %s, got: %s", mvccpb.PUT, watchResp.Events[0].Type)
-		}
-		if string(watchResp.Events[0].Kv.Key) != key {
-			t.Fatalf("unexpected key, expected: %s, got: %s", key, string(watchResp.Events[0].Kv.Key))
-		}
-		if string(watchResp.Events[0].Kv.Value) != v6 {
-			t.Fatalf("unexpected valye, expected: %s, got: %s", v6, string(watchResp.Events[0].Kv.Value))
-		}
+		require.Len(t, watchResp.Events, 2)
+
+		require.Equal(t, mvccpb.DELETE, watchResp.Events[0].Type)
+		deletedKey := string(watchResp.Events[0].Kv.Key)
+		require.Equal(t, key, deletedKey)
+
+		require.Equal(t, mvccpb.PUT, watchResp.Events[1].Type)
+
+		updatedKey := string(watchResp.Events[1].Kv.Key)
+		require.Equal(t, key, updatedKey)
+
+		require.Equal(t, v6, string(watchResp.Events[1].Kv.Value))
 	case <-time.After(100 * time.Millisecond):
 		// we care only about the first response, but have an
 		// escape hatch in case the watch response is delayed.
 		t.Fatal("timed out getting watch response")
 	}
 }
+
+func TestVerifyHashKVAfterCompact(t *testing.T) {
+	tests := []struct {
+		compactedOnTombstoneRev bool
+		hashKVOnCompactedRev    bool
+	}{
+		{
+			compactedOnTombstoneRev: true,
+			hashKVOnCompactedRev:    true,
+		},
+		{
+			compactedOnTombstoneRev: true,
+			hashKVOnCompactedRev:    false,
+		},
+		{
+			compactedOnTombstoneRev: false,
+			hashKVOnCompactedRev:    true,
+		},
+		{
+			compactedOnTombstoneRev: false,
+			hashKVOnCompactedRev:    false,
+		},
+	}
+
+	for idx, tt := range tests {
+		t.Run(fmt.Sprintf("#%d", idx), func(t *testing.T) {
+			verifyHashKVAfterCompactFn(t, e2e.CurrentVersion, func(t *testing.T, cli *e2e.EtcdctlV3, lastRevision int64, tombstoneRevisions []int64) {
+				ctx := context.Background()
+
+				hashKVOnCompactedRev := lastRevision
+				if tt.hashKVOnCompactedRev {
+					hashKVOnCompactedRev = tombstoneRevisions[len(tombstoneRevisions)-1]
+				}
+
+				compactedOnRev := tombstoneRevisions[len(tombstoneRevisions)-1]
+				if !tt.compactedOnTombstoneRev {
+					compactedOnRev = hashKVOnCompactedRev - 8
+				}
+
+				t.Logf("COMPACT rev=%d", compactedOnRev)
+				_, err := cli.Compact(ctx, compactedOnRev, config.CompactOption{Physical: true})
+				require.NoError(t, err)
+
+				t.Logf("HashKV on rev=%d", hashKVOnCompactedRev)
+				resp, err := cli.HashKV(ctx, hashKVOnCompactedRev)
+				require.NoError(t, err)
+
+				require.Len(t, resp, 3)
+				require.True(t, resp[0].Hash != 0)
+				t.Logf("One Hash value is %d", resp[0].Hash)
+
+				require.Equal(t, resp[0].Hash, resp[1].Hash)
+				require.Equal(t, resp[1].Hash, resp[2].Hash)
+			})
+		})
+	}
+}
+
+func TestVerifyHashKVAfterCompactMixVersion(t *testing.T) {
+	if !fileutil.Exist(e2e.BinPath.EtcdLastRelease) {
+		t.Skipf("%q does not exist", e2e.BinPath.EtcdLastRelease)
+	}
+
+	tests := []struct {
+		compactedOnTombstoneRev bool
+		hashKVOnCompactedRev    bool
+		consistentHash          bool
```

Comment:
It's confusing, why do we need this field? All etcd members should always return the same hashKV on the same rev, no matter it's mix-version or not.

Reply:
Please ping me on slack if you want discussion.